driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -638,4 +720,310 @@ public Prune shouldPrune(final UsageTrackingInternalConnection usageTrackingConn
+                        connectionPoolListener, getId(connection));
+                return openAndReleasePermit(connection, asyncCallback);
+            }
+        }
+
+        /**
+         * @return Either {@code null} if a permit has been acquired, or a {@link PooledConnection}
+         * if {@code tryGetAvailable} is {@code true} and an {@linkplain PooledConnection#opened() opened} one becomes available while
+         * waiting for a permit.
+         * @throws MongoTimeoutException If timed out.
+         */
+        @Nullable
+        private PooledConnection acquirePermitOrGetAvailableOpenConnection(final boolean tryGetAvailable, final Timeout timeout)
+                throws MongoTimeoutException {
+            PooledConnection availableConnection = null;
+            tryLock(timeout);
+            try {
+                if (tryGetAvailable) {
+                    addLastDesiredConnectionSlot();
+                }
+                long remainingNanos = timeout.remainingNanosOrInfinite();
+                while (permits == 0) {
+                    availableConnection = tryGetAvailable ? tryGetAndRemoveFirstDesiredConnectionSlot() : null;
+                    if (availableConnection != null) {
+                        break;
+                    }
+                    if (Timeout.expired(remainingNanos)) {
+                        throw createTimeoutException(timeout);
+                    }
+                    remainingNanos = awaitNanos(remainingNanos);
+                }
+                if (availableConnection == null) {
+                    assert permits > 0 : permits;
+                    permits--;
+                } else {
+                    assert availableConnection.opened();
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace(format("Received opened connection [%s] to server %s",
+                                getId(availableConnection), serverId.getAddress()));
+                    }
+                }
+                return availableConnection;
+            } finally {
+                try {
+                    if (tryGetAvailable && availableConnection == null) {//the desired connection slot has not yet been removed
+                        removeLastDesiredConnectionSlot();
+                    }
+                } finally {
+                    lock.unlock();
+                }
+            }
+        }
+
+        private void releasePermit() {
+            lock();
+            try {
+                assert permits < maxPermits : permits;
+                permits++;
+                condition.signal();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        private void addLastDesiredConnectionSlot() {
+            desiredConnectionSlots.addLast(new MutableReference<>());
+        }
+
+        @Nullable
+        private PooledConnection tryGetAndRemoveFirstDesiredConnectionSlot() {
```

Comment:
> Could this method also handle the removal of the previously signalled demand?

This question implies a situation in which a thread may return from `acquirePermitOrGetAvailableOpenConnection` without removing its expressed desire. This situation is incorrect and we should not allow it to happen in the first place, instead of trying to deal with it later. Anyway, I do not think there is a practically useful way to deal with such a situation post factum; it would have involved somehow storing information about the threads that want (or wanted) to get an available connection.

Reply:
> describe the usecase better than describing the mechanism of how its achieved

Done.